py4DSTEM/process/phase/parameter_optimize.py

```diff
@@ -0,0 +1,449 @@
+        afft_params = list(affine_optimization_params.keys())
+        prep_params = list(preprocess_optimization_params.keys())
+        reco_params = list(reconstruct_optimization_params.keys())
+
+        # Construct partial methods to encapsulate the static parameters.
+        # If only ``reconstruct`` has optimization variables, perform
+        # preprocessing now, store the ptycho object, and use dummy
+        # functions instead of the partials
+        if (len(init_params), len(afft_params), len(prep_params)) == (0, 0, 0):
+            affine_preprocessed = AffineTransform(**affine_static_args)
+            init_args = init_static_args.copy()
+            init_args["initial_scan_positions"] = self._get_scan_positions(
+                affine_preprocessed, init_static_args["datacube"]
+            )
+
+            ptycho_preprocessed = cls(**init_args).preprocess(
+                **preprocess_static_args
+            )
+
+            def obj(**kwargs):
+                return ptycho_preprocessed
+
+            def prep(ptycho, **kwargs):
+                return ptycho
+
+        else:
+            obj = partial(cls, **init_static_args)
+            prep = partial(cls.preprocess, **preprocess_static_args)
+
+        affine = partial(AffineTransform, **affine_static_args)
+        recon = partial(cls.reconstruct, **reconstruct_static_args)
+        
+        # Target function for Gaussian process optimization that takes a single
+        # dict of named parameters and returns the ptycho error metric
+        @use_named_args(parameter_list)
+        def f(**kwargs):
+            init_args = {k: kwargs[k] for k in init_params}
+            afft_args = {k: kwargs[k] for k in afft_params}
+            prep_args = {k: kwargs[k] for k in prep_params}
+            reco_args = {k: kwargs[k] for k in reco_params}
+
+            # Create affine transform object
+            tr = affine(**afft_args)
+            # Apply affine transform to pixel grid, using the
+            # calibrations lifted from the dataset
+            dataset = init_static_args["datacube"]
+            init_args["initial_scan_positions"] = self._get_scan_positions(tr, dataset)
+
+            ptycho = obj(**init_args)
+            prep(ptycho, **prep_args)
+            recon(ptycho, **reco_args)
+
+            return np.log(ptycho.error)
+
+        return f
+
+    def _set_optimizer_defaults(self):
+        """
+        Set all of the verbose and plotting to False
+        """
+        self._init_static_args["verbose"] = False
+
+        self._preprocess_static_args["plot_center_of_mass"] = False
+        self._preprocess_static_args["plot_rotation"] = False
+        self._preprocess_static_args["plot_probe_overlaps"] = False
+
+        self._reconstruction_static_args["progress_bar"] = False
+        self._reconstruction_static_args["store_iterations"] = False
+        self._reconstruction_static_args["reset"] = True
```

Comment:
My thinking is that we probably should not spit out plots on every run, since that will undoubtedly crowd up a notebook. It would not make a lot of sense unless we also had it print out the parameters at each iteration as well, which would need some flags or callbacks or something. Probably the best thing for a user to do is to explore the limits of how much defocus they can handle manually before running the optimizer. Alternatively, if you were to come up with some kind of metric that determines if the probe fits well inside the box, we could have it short-circuit and return the max error if the ptycho is run with parameters that can't fit.

Reply:
OK yeah let's add it